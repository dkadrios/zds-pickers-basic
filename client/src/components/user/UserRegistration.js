import React, { Component } from 'react'
import PropTypes from 'prop-types'
import withStyles from '@material-ui/core/styles/withStyles'
import Button from '@material-ui/core/Button'
import DialogTitle from '@material-ui/core/DialogTitle'
import DialogContent from '@material-ui/core/DialogContent'
import DialogContentText from '@material-ui/core/DialogContentText'
import DialogActions from '@material-ui/core/DialogActions'
import SvgIcon from '@material-ui/core/SvgIcon'
import GravatarIcon from '../../images/Gravatar.svg'
import Dialog from '../Dialog'
import UserRegistrationForm from './UserRegistrationForm'

const styles = {
  icon: {
    width: 36,
    height: 36,
    marginRight: 7,
  },
  link: {
    color: 'black',
  },
  gravatarNotice: {
    display: 'flex',
    maxWidth: 330,
  },
}

export const GravatarLink = ({ content }) => (
  <a
    href="https://en.gravatar.com/"
    rel="noopener noreferrer"
    target="_blank"
  >
    {content}
  </a>
)
GravatarLink.propTypes = { content: PropTypes.node.isRequired }

export class UserRegistration extends Component {
  componentDidMount() {
    document.addEventListener('keydown', this.handleKeyDown, false)
  }

  componentWillUnmount() {
    document.removeEventListener('keydown', this.handleKeyDown, false)
  }

  handleKeyDown = (event) => {
    const { ctrlKey, key, metaKey } = event
    const { hideDialog } = this.props

    // Press CTRL+CMD+U to dismiss registration form
    if (key === 'u' && ctrlKey && metaKey) {
      event.preventDefault()
      event.stopImmediatePropagation()
      const productInstance = {
        registrations: [
          {
            active: true,
            firstName: 'Test',
            lastName: 'User',
            email: 'no_one@email.com',
          },
        ],
      }
      // checkedRegistrationAction(productInstance)
      hideDialog()
    }
  }

  render() {
    const { active, classes, hideDialog, registered, serialNumber, submitRegistrationForm } = this.props
    return (
      <Dialog
        disableBackdropClick={!registered}
        disableEscapeKeyDown={!registered}
        onClose={hideDialog}
        open={active}
      >
        <DialogTitle>ZDS Shifter Registration</DialogTitle>
        <DialogContent>
          <DialogContentText>
            Register yourself as the owner of this Shifter
            <br />
            Serial number:
            <strong> {serialNumber}</strong>
          </DialogContentText>

          <UserRegistrationForm />

          <div className={classes.gravatarNotice}>
            <GravatarLink
              content={(
                <SvgIcon className={classes.icon}>
                  <GravatarIcon />
                </SvgIcon>
)}
            />
            <GravatarLink
              className={classes.link}
              content="We use Gravatar for displaying profile images, based on your email address"
            />
          </div>
        </DialogContent>
        <DialogActions>
          {Boolean(registered) && (
            <Button
              onClick={hideDialog}
              tag="btnCancel"
            >
              Cancel
            </Button>
          )}
          <Button
            autoFocus
            color="primary"
            onClick={submitRegistrationForm}
            tag="btnRegister"
            variant="contained"
          >
            Register
          </Button>
        </DialogActions>
      </Dialog>
    )
  }
}

UserRegistration.propTypes = {
  active: PropTypes.bool.isRequired,
  hideDialog: PropTypes.func.isRequired,
  submitRegistrationForm: PropTypes.func.isRequired,
  serialNumber: PropTypes.string.isRequired,
  classes: PropTypes.object.isRequired,
  registered: PropTypes.bool.isRequired,
}

export default withStyles(styles)(UserRegistration)
