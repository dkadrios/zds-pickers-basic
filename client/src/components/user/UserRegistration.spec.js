import set from 'lodash/fp/set'
import { GravatarLink, UserRegistration } from './UserRegistration'

const hideDialog = jest.fn()
const submitRegistrationForm = jest.fn()
const checkedRegistrationAction = jest.fn()

const props = {
  active: true,
  hideDialog,
  submitRegistrationForm,
  checkedRegistrationAction,
  serialNumber: 'serial-num',
  classes: {},
  registered: false,
}

describe('UserRegistration tests', () => {
  beforeEach(jest.clearAllMocks)

  it('renders correctly', () => {
    shallowExpect(UserRegistration)(props).toMatchSnapshot()
  })

  it('renders gravatar correctly', () => {
    shallowExpect(GravatarLink)({ content: 'content' }).toMatchSnapshot()
  })

  it('displays cancel button once registered', () => {
    expect(shallow(createElement(UserRegistration)(props))
      .find('WithStyles(Button)')
      .filter('[tag="btnCancel"]')).toHaveLength(0)

    expect(shallow(createElement(UserRegistration)(set('registered', true)(props)))
      .find('WithStyles(Button)')
      .filter('[tag="btnCancel"]')).toHaveLength(1)
  })
})
