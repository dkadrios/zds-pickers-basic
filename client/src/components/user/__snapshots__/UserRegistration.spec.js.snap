// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`UserRegistration tests renders correctly 1`] = `
<Dialog
  disableBackdropClick={true}
  disableEscapeKeyDown={true}
  onClose={[MockFunction]}
  open={true}
>
  <WithStyles(DialogTitle)>
    ZDS Shifter Registration
  </WithStyles(DialogTitle)>
  <WithStyles(DialogContent)>
    <WithStyles(DialogContentText)>
      Register yourself as the owner of this Shifter
      <br />
      Serial number:
      <strong>
         
        serial-num
      </strong>
    </WithStyles(DialogContentText)>
    <Connect(ReduxForm) />
    <div>
      <GravatarLink
        content={
          <WithStyles(SvgIcon)>
            <test-file-stub />
          </WithStyles(SvgIcon)>
        }
      />
      <GravatarLink
        content="We use Gravatar for displaying profile images, based on your email address"
      />
    </div>
  </WithStyles(DialogContent)>
  <WithStyles(DialogActions)>
    <WithStyles(Button)
      autoFocus={true}
      color="primary"
      onClick={[MockFunction]}
      tag="btnRegister"
      variant="contained"
    >
      Register
    </WithStyles(Button)>
  </WithStyles(DialogActions)>
</Dialog>
`;

exports[`UserRegistration tests renders gravatar correctly 1`] = `
<a
  href="https://en.gravatar.com/"
  rel="noopener noreferrer"
  target="_blank"
>
  content
</a>
`;
